src/lib/alloc.c

```diff
@@ -656,40 +656,32 @@ void *_balloc(int zone, uint32_t caps, size_t bytes)
 	for (i = 0; i < heap->blocks; i++) {
 		map = &heap->map[i];
 
-		/* is block big enough */
-		if (map->block_size < bytes)
-			continue;
-
-		/* does block have free space */
-		if (map->free_count == 0)
-			continue;
-
-		/* allocate block */
-		ptr = alloc_block(heap, i, caps);
-		goto out;
+		/* Check if blocks are big enough and at least one is free */
+		if (map->block_size >= bytes && map->free_count) {
+			/* found: grab a block */
+			ptr = alloc_block(heap, i, caps);
+			break;
+		}
 	}
 
 	/* request spans > 1 block */
-
-	/* only 1 choice for block size */
-	if (heap->blocks == 1) {
-		ptr = alloc_cont_blocks(heap, 0, caps, bytes);
-		goto out;
-	} else {
-
-		/* find best block size for request */
-		for (i = 0; i < heap->blocks; i++) {
+	if (!ptr) {
+		/*
+		 * Find the best block size for request. We know, that we failed
+		 * to find a single large enough block, so, skip those.
+		 */
+		for (i = heap->blocks - 1; i >= 0; i--) {
```

Comment:
Question arises here. Since logic inversion we will use the bigger blocks first. If we leave logic as was

> for (i = 0; i < heap->blocks; i++)

then we will use smaller ones first  for continuous allocation.
@lgirdwood What do You think about it ?

Reply:
I might be missing something, but I don't understand how this loop in the previous version worked at all. In the loop body it had:
```
-                       /* allocate is block size smaller than request */
-                       if (map->block_size < bytes)
-                               alloc_cont_blocks(heap, i, caps, bytes);
```
note - without assigning the result of `alloc_cont_blocks()` to ptr! Wasn't that simply a bug? And the only rescue that actually was saving the code was the line below the loop:
```
-       ptr = alloc_cont_blocks(heap, heap->blocks - 1, caps, bytes);
```

Reply:
...but in fact when I was writing that, I looked at various other block maps, e.g. for runtime zones, which indeed arrange block in the order of increasing size, so, listing backwards we make sure we don't exhaust pools of small buffers for larger allocations. Map arrays for the buffer zone on all architectures only have one entry, so, for them the order is *currently* unimportant. Still, since that increasing order seems to be a common pattern, I think it would be better to use it.